Fixed persistent volume claim controllers processing an old claim.

jsafrane opened the pull request:
Fixes #19860 (it may be easier to look at the issue to see exact sequence
to reproduce the bug and understand the fix).

When `PersistentVolumeProvisionerController.reconcileClaim()` is called with the
same claim in short succession (e.g. the claim is created by an user and
at the same time periodic check of all claims is scheduled), the second
`reconcileClaim()` call gets an old copy of the claim as its parameter.

The method should always reload the claim to get a fresh copy with all
annotations, possibly added by previous `reconcileClaim()` call.

The same applies to `PersistentVolumeClaimBinder.syncClaim()`.

jsafrane commented:
Of course, an unit test is on my TODO list. At least we can talk about the fix.

jsafrane commented:
@k8s-bot test this

jsafrane commented:
Pushed with an unit test. Also, some old tests had to be updated to store claims in a fake API server before calling `reconcileClaim` and `syncClaim` - these functions will reload it from there.

bprashanth commented:
@saad-ali would you like me to do this review? I can, ETA is early next week though

markturansky commented:
thanks @bprashanth but I can take point on this one.

markturansky commented:
@jsafrane what if we removed claim updating from the provisioning entirely?  We can do the PV creation in the binder while still allowing the provisioner to fulfill the volume.  The provisioner is like the recycler in that it really only needs to work on volumes.

jsafrane commented:
> We can do the PV creation in the binder

This leaks bits of the provisioning into binder controller. I am not against though, it would make the provisioner controller simpler, operating on PVs only.

And it still leaves the race in binder controller, this PR is still needed.

markturansky commented on pkg/controller/persistentvolume/persistentvolume_provisioner_controller.go:
```diff
@@ -153,6 +153,16 @@ func (controller *PersistentVolumeProvisionerController) handleUpdateClaim(oldOb
+	if err != nil {
+		return fmt.Errorf("Cannot reload claim %s/%s: %v", claim.Namespace, claim.Name, err)
+	}
+	claim = newClaim
```
per our IRC discussion, please update the local cache with the latest and greatest version.

jsafrane replied:
Is it safe to update `controller.claimStore` here? Won't it generate a new 'claim updated' event?

markturansky replied:
It is safe.  API updates generate events.  The watch gets the events and puts them in the local store.

jsafrane replied:
I added update few lines below.

jsafrane committed:
Fixed persistent volume claim controllers processing an old claim.

Fixes #19860 (it may be easier to look at the issue to see exact sequence
to reproduce the bug and understand the fix).

When PersistentVolumeProvisionerController.reconcileClaim() is called with the
same claim in short succession (e.g. the claim is created by an user and
at the same time periodic check of all claims is scheduled), the second
reconcileClaim() call gets an old copy of the claim as its parameter.

The method should always reload the claim to get a fresh copy with all
annotations, possibly added by previous reconcileClaim() call.


The same applies to PersistentVolumeClaimBinder.syncClaim().


Also update all the test to store claims in "fake" API server before calling
syncClaim and reconcileClaim.

jsafrane commented:
There is something wrong with this PR that breaks integration tests. I am quite  sure it's not a flake, it always fails with the patch and always succeeds without it.

```
I0201 14:22:49.525423    4919 trace.go:57] Trace "Create /api/v1/namespaces/default/persistentvolumeclaims" (started 2016-02-01 14:22:49.274322577 +0000 UTC):
[18.296Âµs] [18.296Âµs] About to convert to expected version
[99.985Âµs] [81.689Âµs] Conversion done
[103.4Âµs] [3.415Âµs] About to store object in database
[250.913786ms] [250.810386ms] Object stored in database
[250.931998ms] [18.212Âµs] Self-link added
[251.068318ms] [136.32Âµs] END
panic: test timed out after 5m0s

<lot of stack traces>
...
FAIL	k8s.io/kubernetes/test/integration	300.037s
```

@markturansky, any idea where to look?

jsafrane commented:
> There is something wrong with this PR that breaks integration tests.

The integration tests use a kube client with default QPS and Burst values and at the same time it runs periodical update of binder controller every 10 seconds. Since the test creates 100 claims and the binder controller re-reads the claims from etcd, the client slows down the requests. And after 10 seconds a periodic check kicks in, making things even worse...

I updated QPS and Burst values of the test client to make sure it does not start throttling the requests.

markturansky commented:
Try reducing the claim count to 5-10 in rapid succession?  Making 100 is a bit of a ridiculous test (I know I added it).

jsafrane commented:
> Try reducing the claim count to 5-10 in rapid succession? Making 100 is a bit of a ridiculous test (I know I added it).


The test as patched in this PR works. Is reducing nr. of claims better than increasing burst & QPS? I have no opinion here, I'm just asking what's better.

markturansky commented:
I have no preference.  Reducing the number of claims from 100 to 10 still tests the concurrency of the controller. I'm ok with that.  Whichever is easier, frankly.

jsafrane commented:
Ok, I would like to leave the patch as it is, with higher QPS. If I am not mistaken, all your review remarks was adressed.

markturansky commented:
LGTM.   

@swagiaal 2nd opinion?

This PR might fix some needed issues today, but I think we're all realizing now how a better controller design will prevent the need for something like this.

swagiaal commented:
So I think this narrows the window for the race but does not eliminate it :/
Hopefully future work on using resource versions to ensure congruence will alleviate the need for this. This LGTM.
@markturansky back to you

jsafrane commented:
@k8s-bot test this issue: #21045
